near_api/providers.py

```diff
@@ -18,28 +23,104 @@ class FinalityTypes:
 
 
 class JsonProviderError(Exception):
-    pass
+    def get_type(self) -> Optional[str]:
+        try:
+            return self.args[0]["name"]
+        except (IndexError, KeyError):
+            return None
+
+    def get_cause(self) -> Optional[str]:
+        try:
+            return self.args[0]["cause"]["name"]
+        except (IndexError, KeyError):
+            return None
+
+    def is_invalid_nonce_tx_error(self) -> bool:
+        try:
+            return (
+                self.get_type() == "HANDLER_ERROR"
+                and self.get_cause() == "INVALID_TRANSACTION"
+                and "InvalidNonce" in self.args[0]["data"]["TxExecutionError"]["InvalidTxError"]
+            )
+        except (IndexError, KeyError):
+            return False
+
+    def is_expired_tx_error(self) -> bool:
+        try:
+            return (
+                self.get_type() == "HANDLER_ERROR"
+                and self.get_cause() == "INVALID_TRANSACTION"
+                and self.args[0]["data"]["TxExecutionError"]["InvalidTxError"] == "Expired"
+            )
+        except (IndexError, KeyError):
+            return False
 
 
 class JsonProvider(object):
-    def __init__(self, rpc_addr, proxies=None):
+    def __init__(
+            self,
+            rpc_addr,
+            proxies=None,
+            tx_timeout_retry_number: int = 12,
+            tx_timeout_retry_backoff_factor: float = 1.5,
+            http_retry_number: int = 10,
+            http_retry_backoff_factor: float = 1.5,
+            session: Optional[requests.Session] = None,
+    ) -> None:
         if isinstance(rpc_addr, tuple):
             self._rpc_addr = "http://%s:%s" % rpc_addr
         else:
             self._rpc_addr = rpc_addr
         self.proxies = proxies
 
+        if session is None:
+            # Loosely based on https://findwork.dev/blog/advanced-usage-python-requests-timeouts-retries-hooks/
+            adapter = requests.adapters.HTTPAdapter(
+                max_retries=urllib3.util.Retry(
+                    total=http_retry_number,
+                    backoff_factor=http_retry_backoff_factor,
+                    status_forcelist=[502, 503],
+                    allowed_methods=["GET", "POST"],
+                ),
+            )
+
+            session = requests.Session()
+            session.mount("https://", adapter)
+            session.mount("http://", adapter)
```

Comment:
I used sessions instead of manual retrying because a built-in mechanism seem more straightforward to me and it allows great customisation (user can pass their own Session object).

If you prefer I can change it to manual exception catching and handling.

Reply:
This is a bit specious to me.  There are now three nested retry loops.  We have

1. a loop in _sign_and_submit_tx which retries up to 12 times if we had wrong nonce or block,
2. another loop here in json_rpc which retries up to 10 times if we had a timeout and
3. final loop in the Retry adapter which retries up to 10 times on server errors.

It may be somewhat contrived, but in pathological case we may end up retrying 1200 times.  I’d say we should at least get rid of the Retry adapter and to the looping logic for 5xx errors in json_rpc.  Something like:

```python
        while True:
            try:
                return self._json_rpc_once(method, params, timeout)
            except HTTPError as e:  # whatever the exception type is
                if attempt >= self._tx_timeout_retry_number:
                    raise
                if e.status_code // 100 == 5:
                    log.warning("Retrying request to %s as it returned server error: %s", method, e)
                else:
                    raise
            except JsonProviderError as e:
                if attempt >= self._tx_timeout_retry_number:
                    raise
                if e.get_type() == "HANDLER_ERROR" and e.get_cause() == "TIMEOUT_ERROR":
                    log.warning("Retrying request to %s as it has timed out: %s", method, e)
                else:
                    raise
            attempt += 1
            time.sleep(self._tx_timeout_retry_backoff_factor ** attempt * 0.5)
```

Reply:
Yeah, that's what is done by near-api-js ;p

Ok, I'll refactor this to have only two levels of retries

Reply:
Yeah, I’ve noticed.  I’d argue it’s a bug in JS implementation.